pkg/applicationserver/applicationserver.go

```diff
@@ -512,35 +513,46 @@ func (d downlinkQueueOperation) shouldSkip(sessionKeyID []byte) bool {
 
 const maxDownlinkQueueOperationAttempts = 50
 
-func (as *ApplicationServer) attemptDownlinkQueueOp(ctx context.Context, dev *ttnpb.EndDevice, link *ttnpb.ApplicationLink, peer cluster.Peer, op downlinkQueueOperation) ([]string, error) {
+func (as *ApplicationServer) attemptDownlinkQueueOp(ctx context.Context, dev *ttnpb.EndDevice, link *ttnpb.ApplicationLink, peer cluster.Peer, op downlinkQueueOperation) (mask []string, err error) {
+	resultCallbackItems := op.Items
+	if op.ResultFunc != nil {
+		defer func() {
+			op.ResultFunc(resultCallbackItems, err)
+		}()
+	}
+
+	sessions := make([]*ttnpb.Session, 0, 2)
+	if dev.Session != nil && !op.shouldSkip(dev.Session.Keys.SessionKeyId) {
+		sessions = append(sessions, dev.Session)
+		mask = ttnpb.AddFields(mask, "session.last_a_f_cnt_down")
+	}
+	if dev.PendingSession != nil && !op.shouldSkip(dev.PendingSession.Keys.SessionKeyId) {
+		// Downlink can be encrypted with the pending session while the device first joined but not confirmed the
+		// session by sending an uplink.
+		sessions = append(sessions, dev.PendingSession)
+		mask = ttnpb.AddFields(mask, "pending_session.last_a_f_cnt_down")
+	}
+	if len(sessions) == 0 {
+		return nil, errNoDeviceSession.New()
+	}
+
+	encryptedItems, err := as.encodeAndEncryptDownlinks(ctx, dev, link, op.Items, sessions)
+	if err != nil {
+		return nil, err
+	}
+	// When encoding and encryption succeeds, pass the encoded and encrypted items to the result callback. This is useful
+	// for diagnostics as error or successful forwarding events contains the actual FRMPayload and FPort and not the
+	// original input.
+	resultCallbackItems = encryptedItems
```

Comment:
Are you really sure this does what it says ? `encodeAndEncryptDownlinks` erases the decoded payload.